src/lib/FreeC/Pass/KindCheckPass.hs

```diff
@@ -0,0 +1,180 @@
+
+-- | This module contains a compiler pass that checks if all type constructors
+--   in type expressions are fully applied and type variables are not applied.
+--
+--   For simplification, all type variables have kind @*@. Therefore, all type
+--   constructors must be fully applied, so partial application of type
+--   constructors is not allowed. This pass also checks that type variables and
+--   function types do not occur on the left hand side of a type application.
+--   If a type expression is found that does not match this rules, a fatal
+--   error is reported. For simplification, a type expression that satisfies all
+--   of these rules is called kind-correct.
+--
+--   = Example
+--
+--   The following module contains three invalid definitions.
+--
+--   > module Failures where
+--   >
+--   > data MyList a = MyNil | MyCons a (MyList a)
+--   >
+--   > -- Invalid because MyList is applied to 2 arguments
```

Comment:
Comments should be full sentences. End them with a period (also check the other comments in this file). 

```suggestion
--   > -- Invalid because 'MyList' is applied to 2 arguments.
```

Furthermore, I'd use Haddock notation in the examples. In this particular case the reader can understand what `MyList` is even without markup. However, in the comment for `fail2` it is not entirely clear whether `a` is the variable name or a misplaced article.